app/src/main/java/org/thoughtcrime/securesms/sms/IncomingTextMessage.java

```diff
@@ -0,0 +1,314 @@
+    this.replyPathPresent        = message.isReplyPathPresent();
+    this.pseudoSubject           = message.getPseudoSubject();
+    this.sentTimestampMillis     = message.getTimestampMillis();
+    this.serverTimestampMillis   = -1;
+    this.receivedTimestampMillis = System.currentTimeMillis();
+    this.subscriptionId          = subscriptionId;
+    this.expiresInMillis         = 0;
+    this.groupId                 = null;
+    this.push                    = false;
+    this.unidentified            = false;
+    this.serverGuid              = null;
+  }
+
+  public IncomingTextMessage(@NonNull RecipientId sender,
+                             int senderDeviceId,
+                             long sentTimestampMillis,
+                             long serverTimestampMillis,
+                             long receivedTimestampMillis,
+                             String encodedBody,
+                             Optional<GroupId> groupId,
+                             long expiresInMillis,
+                             boolean unidentified,
+                             String serverGuid)
+  {
+    this.message                 = encodedBody;
+    this.sender                  = sender;
+    this.senderDeviceId          = senderDeviceId;
+    this.protocol                = 31337;
+    this.serviceCenterAddress    = "GCM";
+    this.replyPathPresent        = true;
+    this.pseudoSubject           = "";
+    this.sentTimestampMillis     = sentTimestampMillis;
+    this.serverTimestampMillis   = serverTimestampMillis;
+    this.receivedTimestampMillis = receivedTimestampMillis;
+    this.push                    = true;
+    this.subscriptionId          = -1;
+    this.expiresInMillis         = expiresInMillis;
+    this.unidentified            = unidentified;
+    this.groupId                 = groupId.orNull();
+    this.serverGuid              = serverGuid;
+  }
+
+  public IncomingTextMessage(Parcel in) {
+    this.message                 = in.readString();
+    this.sender                  = in.readParcelable(IncomingTextMessage.class.getClassLoader());
+    this.senderDeviceId          = in.readInt();
+    this.protocol                = in.readInt();
+    this.serviceCenterAddress    = in.readString();
+    this.replyPathPresent        = (in.readInt() == 1);
+    this.pseudoSubject           = in.readString();
+    this.sentTimestampMillis     = in.readLong();
+    this.serverTimestampMillis   = in.readLong();
+    this.receivedTimestampMillis = in.readLong();
+    this.groupId                 = GroupId.parseNullableOrThrow(in.readString());
+    this.push                    = (in.readInt() == 1);
+    this.subscriptionId          = in.readInt();
+    this.expiresInMillis         = in.readLong();
+    this.unidentified            = in.readInt() == 1;
+    this.serverGuid              = in.readString();
+  }
+
+  public IncomingTextMessage(IncomingTextMessage base, String newBody) {
+    this.message                 = newBody;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;newBody&#39; (index: 1 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;base&#39;. The same parameter type is validated here for example: app/src/flipper/java/org/thoughtcrime/securesms/database/FlipperSqlCipherAdapter.java:84. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).